### PR TITLE
🐛 ClusterResourceSet controller: fix reconciliation triggered by resource changes

### LINKED ISSUE #4486: Periodic tests are failing: exp/addons/controllers.TestAPIs
**What steps did you take and what happened:**
[A clear and concise description on how to REPRODUCE the bug.]

According to [testgrid](https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-test-main&width=20&sort-by-failures=&exclude-non-failed-tests=100) the TestAPIs test is regularly failing in our periodic tests of the main branch (rough estimate 10-20%)

image

/kind failing-test

/cc @fabriziopandini

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
The ClusterResourceSet controller did not react to changes on
ConfigMaps and Secrets (except when OwnerReferences were already set).

The problem was that we use OnlyMetadata on the watches and we tried to
get the kind of ConfigMap/Secret from the TypeMeta of an PartialObjectMetadata.
The TypeMeta of a PartialObjectMetadata is empty so apiutil.GVKForObject(o,
r.Client.Scheme()) always returned an error and the reconciliation was
not triggered.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4486

Additional context:
* https://kubernetes.slack.com/archives/C8TSNPY4T/p1618606088075900
* branch to reproduce the issue easily: https://github.com/kubernetes-sigs/cluster-api/pull/4490
  * explanations on how to do that are in the Slack thread

### CODE REVIEW

File: exp/addons/controllers/clusterresourceset_controller_test.go [hunk 1]
```diff
@@ -292,21 +292,18 @@ metadata:
...
+		// of another resource triggers the reconciliation and we cannot verify if we react
+		// correctly to the ConfigMap creation
+		time.Sleep(5 * time.Second)
+
 		Expect(testEnv.Create(ctx, newConfigmap)).To(Succeed())
 		defer func() {
 			Expect(testEnv.Delete(ctx, newConfigmap)).To(Succeed())
 		}()
 
-		cmKey := client.ObjectKey{
```
Comment (sbueringer) [inline exp/addons/controllers/clusterresourceset_controller_test.go:300]: I'm not sure what the purpose of the Get was. It should just return instantly because the Create above is synchronous. Maybe it was a try to ensure that the ConfigMap exists in the local cache to fix some test flakiness?

File: exp/addons/controllers/clusterresourceset_controller_test.go [hunk 2]
```diff
@@ -292,21 +292,18 @@ metadata:
 			},
 			Data: map[string]string{},
 		}
```
Comment (sbueringer) [inline exp/addons/controllers/clusterresourceset_controller_test.go:294]: We're already trying to wait until a first reconciliation is done in l.276-l.286. Unfortunately this doesn't seem to be enough. If there are other/better ways to wait until everything settles, please let me know.

File: exp/addons/controllers/clusterresourceset_controller_test.go [hunk 3]
```diff
@@ -292,21 +292,18 @@ metadata:
...
+
+		// Let's wait until the initial reconciliations are *all* done, so we can test
+		// if the ConfigMap creation triggers a reconciliation. Otherwise a previous change
+		// of another resource triggers the reconciliation and we cannot verify if we react
+		// correctly to the ConfigMap creation
+		time.Sleep(5 * time.Second)
```
Comment (sbueringer) [inline exp/addons/controllers/clusterresourceset_controller_test.go:300]: This has been added to ensure the test would consistently fail without the fix in the ClusterRersourceSet controller.

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 1 above)
Comment (JoelSpeed) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: > Maybe it was a try to ensure that the ConfigMap exists in the local cache to fix some test flakiness?

This sounds about right 😅  All the controller runtime stuff uses cached clients normally so if `testEnv` is also caching, this would be it waiting until the informer has received the create event and updated the cache. Looking at the next `Eventually`, this basically allows us to know that the timeout is used solely for that reconcile. 

Have you tried running this many times to see if you can get it to fail without this extra block?

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 3 above)
Comment (JoelSpeed) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: Short of wrapping the reconciler to count reconciles (which I would recommend avoiding), I don't think there's really anything that can be done. You also need to account for other controllers and tests in the test suite that may also be creating events, so to get the queue to truly be empty will be very very difficult AFAIK.

If we wanted to be more certain that, we could potentially create a new test with a separate manager watching a different test namespace (I'm not sure this is actually possible here, this is more general advice), that would then be more isolated. In these scenarios it's easier to verify reaction to specific events occurring.

Out of interest, without this, were you seeing the test flaking at all?

Comment (enxebre) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: not ideal but could think about something like adding an optional event recorder to the `resourceToClusterResourceSet` signature that we only pass through in tests that would record the exact resource and operation triggering the reconciliation

Comment (fabriziopandini) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: What about relying on Observed generation instead?

**Review by fabriziopandini — COMMENTED** [review]
@sbueringer thanks for the huge work in investigating this issue.
Before green lighting the fix I would like to understand with @vincepri if there is chance/It should be better to fix this tricky error up in the stack (controller runtime or client-go)

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 1 above)
Comment (sbueringer) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: Not yet, I'll try. Although the 10s timeout is easily enough. So it's highly likely that there won't be any difference.

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 3 above)
Comment (sbueringer) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: @JoelSpeed 
The problem is that without the timeout the test was not reliably able to detect the broken code. So this means when I drop the timeout the test has a certain probability that it doesn't test the reconcile of the create ConfigMap but one of the previous one either Cluster or ClusterResourceSet creation. So this only makes the test "greener" but just because we're testing the wrong thing. So the test was flaky because it couldn't reliably detect the broken code, with the timeout it can.

@enxebre Would it be possible with this solution to detect which exact reconcile of which event lead to the condition we're checking in l.313? If it was already done by a previous reconcile and the "right" reconcile does a no-op, the test is already testing the wrong thing.

@fabriziopandini Do you mean instead of the sleep I should wait until the resources has a specific ObservedGeneration and only after that create the ConfigMap? I'm not entirely sure which reconcilation is currently the problem. I suspect there might be one which is actually a no-op. I'm not sure if I can check for that one when it does nothing :/.

**Comment by sbueringer** [comment]
> @sbueringer thanks for the huge work in investigating this issue.
> Before green lighting the fix I would like to understand with @vincepri if there is chance/It should be better to fix this tricky error up in the stack (controller runtime or client-go)

Yeah of course, no problem. My take is that we cannot fix it upstream, because the `builder.OnlyMetadata` watch behaves as intended. It was a misunderstanding the `PartialObjectMetadata` contains the `TypeMeta` of our `ConfigMap`, as it has to contain the `TypeMeta` of the `PartialObjectMetadata` first.

But of course there might be entirely different ways to fix this.

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 3 above)
Comment (sbueringer) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: @fabriziopandini I added the following here: https://github.com/kubernetes-sigs/cluster-api/blob/2ff65de19a19e018c5ba7fb75376a858861fc625/exp/addons/controllers/clusterresourceset_controller.go#L117
```golang
fmt.Printf("reconcile: %v %v\n", req.NamespacedName, clusterResourceSet.ObjectMeta.Generation)
```
output: (all before `testEnv.Create`):
```bash
reconcile: default/clusterresourceset-pug52m 1
reconcile: default/clusterresourceset-pug52m 1
reconcile: default/clusterresourceset-pug52m 1
reconcile: default/clusterresourceset-pug52m 2
```

So think you're right. I can just wait until `observedGeneration`: `2` instead of the sleep. I was suspecting the last reconcile triggers another one which doesn't change the resource, but apparently I was wrong.

WDYT?

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 1 above)
Comment (sbueringer) [reply exp/addons/controllers/clusterresourceset_controller_test.go:300]: I ran it locally in a loop a few hundred times. But I think we don't need the `Get`. The `Eventually` below tries for 20 seconds and locally it takes `< 1s`. But I don't really care, if the consensus is that we should keep it, I'll just add it again.

**Comment by vincepri** [comment]
/hold

I might have a solution for this in Controller Runtime, which is due to review https://github.com/kubernetes-sigs/controller-runtime/pull/1484

**Comment by sbueringer** [comment]
> /hold
> 
> I might have a solution for this in Controller Runtime, which is due to review [kubernetes-sigs/controller-runtime#1484](https://github.com/kubernetes-sigs/controller-runtime/pull/1484)

Sounds reasonable, essentially looks like the same fix just one layer below. Didn't thought of that. 

If that one is merged, I would drop the changes to the controller and only fix the test.

**Comment by sbueringer** [comment]
I rebased the PR and dropped the changes to the controller as the upgrade to the new controller-runtime version should fix the issue. 

I think it still makes sense to improve the test so it actually tests correct (i.e. it tests the reconciliation triggered by the ConfigMap creation).

I'm not sure what we should do about: https://github.com/kubernetes-sigs/cluster-api/pull/4491#discussion_r615144616

I still don't have a better solution than adding the sleep.

**Comment by sbueringer** [comment]
@vincepri @fabriziopandini any ideas how/ if we should continue here?

**Comment by sbueringer** [comment]
(Had to rebase because of the test refactoring, but essentially same situation as before)

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 3 above)
Comment (fabriziopandini) [reply exp/addons/controllers/clusterresourceset_controller_test.go:312]: @sbueringer 
Catching up after lagging a little bit on this...

I think we should avoid to add a sleep here, because ensuring that watches with partial metadata works as expected is not really something a behavioural test for ClusterResourceSet should be concerned of.

If we want to have test coverage ensures that watches with partial metadata works well, I think we should have a dedicated test, might be under `util/predicates` so we can separate this issue from the CRS specific tests. 
This will allow us to use a dedicated, super simple, test reconciler that e.g. increases an annotation value at every reconcile, so we can actually check when a reconcile happens.

Does this makes sense? This can be eventually addressed also in a separated PR...

File: exp/addons/controllers/clusterresourceset_controller_test.go (hunk 1 above)
Comment (vincepri) [reply exp/addons/controllers/clusterresourceset_controller_test.go:312]: Let's add it back, having a `Get` to make sure the cache syncs is the most idiomatic way to deal with caches and controllers lagging behind a bit

**Comment by vincepri** [comment]
Should we close this?

**Comment by sedefsavas** [comment]
Created a new PR that solves this issue: https://github.com/kubernetes-sigs/cluster-api/pull/4723

**Comment by sbueringer** [comment]
New PR looks good to me
